Textrender Refactoring and Improvements

TsFreddie opened the pull request:
## A mega PR regarding Textrender

This PR does ~~three~~ four things:
- Refactor `text.cpp`
- Reduce stutter when rendering glyphs
- Add multi-font fallback support as well as regional variant switching
- Improve the feature and speed of textrender by reworking the "text layout engine" part and its API (new)

Notable changes:

- Implemented an atlas packing algorithm to tightly pack glyphs onto a fixed sized texture
  - There is no elegant way to resize a texture without re-rendering all glyphs. 
  - Might as well remove texture resizing all together.
- Removes ClFontfile in favour of a json index
- Added Source Han Sans as the fallback font
- Fonts are loaded as OpenType Collection by default.
- Fonts are referenced by font family names and style names instead of filenames. (see datasrc/fonts/index.json)

Fallback:

- Font fallbacks are defined in three stages and checked in the following order:
  1. A default font (Deja Vu)
  2. A regional variant font (Source Han Sans SC for Simplified Chinese etc)
  3. Multiple fallback fonts (Source Han Sans for rest of the world, and maybe icon fonts in the future)

Cons:
- Does not support fonts with no charset map. (Most modern fonts do anyway)

Pros:
- All fonts are stored in the same set of textures

Regional Variants:
- CJK region have different glyphs for displaying the same character sometimes. This is solved by defining a regional variant font.
![image](https://user-images.githubusercontent.com/3797859/92112949-7a663100-ede6-11ea-9164-46f986e18d67.png)


Atlas Packing:
- The texture is partitioned into multiple pages.
- Each page is an atlas texture.
- If the texture is full, drop the page containing the least characters that are shown last frame.

Texture example (1024x1024 texture with 4 pages, the top left page was replaced):
![image](https://user-images.githubusercontent.com/3797859/92098384-0f5e2f80-edd1-11ea-893e-3d8a3ac4f7c5.png)

Optimization:
 - Other than eliminating texture resizing, the old linear glyph finding is replaced with a binary search (using sorted_array).

Word wrapping:
 - Fixed the line wrapping issue by taking utf8 into account when counting word length and check glyphs instead of chars when newlining.
![image](https://user-images.githubusercontent.com/3797859/92179734-33575a80-ee3d-11ea-82b4-2a5f2efe430c.png)

TsFreddie commented:
I want to submit this PR for review first.
I'll try optimizing glyph finding and rendering next to make the large font stutter less.

Robyt3 commented:
Nice work! I think @axblk is already working on optimizing the text rendering.

Check your curly bracket-style. Some lines are not using Allman style.

TsFreddie commented:
Thanks, i'll check that. Thought I already try to avoid my habits. is there a formatter or checker you recommend?

TsFreddie commented:
Now font's style name is considered as well.

- "Font Awesome 5 Free" will match the first face with the name.
- "Font Awesome 5 Free Regular" will match the only face with the name and style.

TsFreddie commented:
Sorry for the title change. I've packed glyphs into atlases and removed the texture resizing. So this PR is doing more than it was originally intended to do.

I might went a bit far with this PR. I can redo some of the features without refactoring if that's preferable.

Robyt3 commented on src/engine/client/textrender.h:
```diff
@@ -0,0 +1,207 @@
+	friend bool operator >=(const CGlyph& l, const CGlyph& r) { return !(l < r); };
+};
+
+class CAtlas
```
Some indentation is inconsistent in this file, it should be all tabs.

Piepow commented on datasrc/fonts/index.json:
```diff
@@ -0,0 +1,16 @@
+        "DejaVuSans.ttf",
+        "SourceHanSans.ttc"
+    ],
+    "default": "DejaVu Sans",
```
This value seems to be ignored. Instead it seems to always take the first font in `font files`

Piepow commented:
I tried testing with some custom fonts (e.g. Roboto, Noto, Tillium Web) and they seem to be working well except for some missing icons :tada:

ChillerDragon commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1169 @@
+
+	int FitWidth = Width;
+
+	if (CurX + Width > m_Width - 1) return -1;
```
while you are at it you could update all these ``if (`` and ``for (`` to ``if(`` and ``for(``

TsFreddie replied:
Just in time before I made a commit. Changed.

ChillerDragon commented on src/engine/client/textrender.h:
```diff
@@ -0,0 +1,207 @@
+	float TextGetLineBaseY(const CTextCursor *pCursor);
+};
+
+#endif
```
New line at the end of file would be nice

TsFreddie commented:
> I tried testing with some custom fonts (e.g. Roboto, Noto, Tillium Web) and they seem to be working well except for some missing icons 🎉

We should design a dedicated icon font and set it as the last fallback, so people who decided to customize their font can still have icons. 

I think it can be useful to place individual svgs files into `datasrc` and let build system build the icon font so we can keep the source of icons not just a font file. But I haven't found any simple and lightweight solution to make that happens.

TsFreddie commented:
The new TextRender API is finished.

The new API avoided a lot of cases where multiple passes are called to calculate text width before rendering it, meanwhile providing built-in text alignment and auto ellipsis. 

The text layout code is rewritten and should be a lot faster. The word wrapping behaviour is changed to a greedy solution which can behave differently than before, but it should be much faster.

This is a huge change to review; the main review focus should be inside `textrender.cpp`. All other changes are made to utilize the new API. For people who are interested in testing, look for any changes in the appearance because the new textrender should look almost identical with the old one.

`chat.cpp` and `console.cpp` can be further optimized, which should be another refactoring job so I kept their logic untouched, which is still pretty slow (but faster during text call nonetheless).

Dune-jr commented:
That looks awesome, especially with the instream of CJK players :)

Robyt3 commented on src/engine/client/textrender.h:
```diff
@@ -0,0 +1,232 @@
+};
+
+// TODO: use SDF or MSDF font instead of multiple font sizes
+static int aFontSizes[] = {8,9,10,11,12,13,14,15,16,17,18,19,20,36,64};
```
s_aFontSizes

Robyt3 commented on src/engine/client/textrender.h:
```diff
@@ -0,0 +1,232 @@
+	float m_BearingX;
+	float m_BearingY;
+	float m_AdvanceX;
+	float m_aUvs[4];
```
m_aUvs could be named more descriptively.

TsFreddie replied:
since it is just UV coordinates. how about `m_aUvCoords`

Robyt3 replied:
If that's correct, go for it. I assumed it was something like a UV scale.

Robyt3 commented on src/engine/client/textrender.h:
```diff
@@ -0,0 +1,232 @@
+
+	bool isWestern(int Chr)
+	{
+		if(Chr >= 0x0020 && Chr <= 0x218F)
```
Just `return Chr >= 0x0020 && Chr <= 0x218F`

The function could be declared `const` for good measure.

Robyt3 commented on src/engine/client/textrender.h:
```diff
@@ -0,0 +1,232 @@
+	void LoadFonts(IStorage *pStorage, IConsole *pConsole);
+	void SetFontLanguageVariant(const char *pLanguageFile);
+
+	// void SetCursor(CTextCursor *pCursor, float x, float y, int Flags = 0);
```
Remove

Robyt3 commented on src/game/client/components/chat.cpp:
```diff
@@ -1309,16 +1314,28 @@ void CChat::OnRender()
-			TextRender()->TextColor(TextColor);
-			TextRender()->TextOutlineColor(ColorHighlightOutline);
-			TextRender()->TextEx(&Cursor, pLine->m_aText, -1);
+			TextRender()->TextColor(TextColor.r, TextColor.g, TextColor.b, TextColor.a);
```
The other TextColor and TextSecondaryColor methods can be used to to pass the vec4 directly.

The call to TextColor appears first in both if branches, so it can be pulled up.

axblk commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1199 @@
+
+float CTextRender::TextWidth(float FontSize, const char *pText, int Length)
+{
+	CTextCursor Cursor(FontSize, 0, 0, TEXTFLAG_NO_RENDER);
```
static

axblk commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1205 @@
+	int W = m_aAtlasPages[Atlas].GetWidth();
+	int H = m_aAtlasPages[Atlas].GetHeight();
+
+	unsigned char *pMem = (unsigned char *)mem_alloc(W*H, 1);
```
memory is not freed afterwards

Robyt3 commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1205 @@
+	mem_free(pMem);
+}
+
+int CGlyphMap::FitGlyph(int Width, int Height, ivec2 *Position)
```
pPosition

Robyt3 commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1207 @@
+
+	int MaxLines = pCursor->m_MaxLines;
+	if(MaxLines < 0)
+		MaxLines = INT32_MAX;
```
`INT32_MAX` is not defined on windows. Should be `INT_MAX`.

Robyt3 commented:
While this improves the shape of glyph outlines, it seems to make some of them darker than they should be. The position of text also seems to change slightly.

Before:

![screenshot_2020-10-04_16-37-06](https://user-images.githubusercontent.com/23437060/95018491-6705ea00-0660-11eb-8f72-97f66a4de788.png)

After:

![screenshot_2020-10-04_16-35-22](https://user-images.githubusercontent.com/23437060/95018486-60777280-0660-11eb-8125-8927f09062f9.png)

TsFreddie commented:
> it seems to make some of them darker than they should be

I removed the multiplication of outline alpha and text alpha when drawing (in favour of a separate alpha value) because the shadow text was doing it that way in master. I tried to apply the separate alpha in scoreboard, it seems more difficult than using the old method, so I pushed a new commit to restore this behaviour based on which texture is being drawn.

> The position of text also seems to change slightly.

For centre / right / middle / bottom aligned text this is expected, since the bounding box calculation is quite different from the old text width / text height calculation.
For vertical position, might be something to do with the pixel alignment code being wrong. I'm Investigating.

TsFreddie commented:
Did some pixel peeping. The pixel alignment is correct. But I did end up realign centre / right aligned text, they may still locate differently than the original code, but at least they are now crystal clear.

Vertical align are now looks the same as original with outline thickness taken into account when calculating glyphs offset.

Also did a rebase to keep up to date.

Robyt3 commented:
Nice.

You also need to adjust

https://github.com/teeworlds/teeworlds/blob/32b820a1d95fc1406b845989414659c7331af553/bam.lua#L388

so the `.ttc` font file is copied from datasrc to the build folder when using bam.

TsFreddie commented:
Another rebases to adapt the new CBroadcast.

In addition:
Added a 24pt fontsize between 20pt and 36pt, reason:
![image](https://user-images.githubusercontent.com/3797859/96402004-701dbc00-1207-11eb-94c2-bf6ddcad0921.png)
In some resolution, the top line is forced to use 36pt font, which is too large and the outline in comparison become too thin. The new atlas packer should be efficient enough to handle a new fontsize.

Changed the font texture to 2048x2048 with 16 pages of atlas, reason:
https://feedback.wildfiregames.com/report/opengl/feature/GL_MAX_TEXTURE_SIZE
Only a few software renderers does not support 2k texture, which makes a 2k default texture relatively safe while having virtually no atlas page drop.

Also removed some redundant data in CTextCursor (namely, `m_GlyphCount` and space glyphs that replace `\n`s even when newlines are allowed).

Finally, we should start considering merging this. There are some works (for example, vote menus, chat optimization, ui text alignments) should come after this, and rebasing for each change related to UI/text is getting increasingly difficult.

Let's get a final round of review. Let's go~ :P

oy commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1227 @@
+		}
+	}
+	m_ActiveAtlasIndex = 0;
+	m_Glyphs.clear();
```
memory leak. doesn't delete the allocated glyphs

oy commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1227 @@
+bool CGlyphMap::SetFaceByName(FT_Face *pFace, const char *pFamilyName)
+{
+	FT_Face Face = NULL;
+	char aFamilyStyleName[128];
```
use a constant

oy commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1227 @@
+	{
+		for(int i = 0; i < m_NumFtFaces; ++i)
+		{
+			str_format(aFamilyStyleName, 128, "%s %s", m_aFtFaces[i]->family_name, m_aFtFaces[i]->style_name);
```
same

TsFreddie commented on src/game/client/components/chat.cpp:
```diff
@@ -756,8 +756,6 @@ void CChat::OnRender()
 {
 	if(Client()->State() < IClient::STATE_ONLINE)
 		return;
-	if(!Config()->m_ClShowChat)
```
mistake when i added chat back probably.

oy commented on src/engine/client/textrender.cpp:
```diff
@@ -0,0 +1,1239 @@
+		if(FT_New_Memory_Face(m_FTLibrary, (FT_Byte *)pBuf, FileSize, i, &FtFace))
+			break;
+
+		if(m_pGlyphMap->AddFace(FtFace))
```
free the face in case it couldn't be added
